docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+     dropped even if the connection remains open.
+
+   - Either the client or the server may close the websocket if the websocket
+     PING and PONG exchanges are not handled correctly, or frequently enough.
+     Even if correctly implemented, this may fail if the system is under high
+     load and cannot service those control messages in a timely manner.
+
+   When the connection is terminated, the server drops all the subscriptions
+   for that client (as if it had called `unsubscribe_all`). Even if the client
+   reconnects, any events that were published during the period between the
+   disconnect and re-connect and re-subscription will be silently lost, and the
+   client has no way to discover that it missed some relevant messages.
+
+3. **No way to replay old events:** Even if a client knew it had missed some
+   events (due to a disconnection, for example), the API provides no way for
+   the client to "play back" events it may have missed.
+
+4. **Large response sizes:** Some event data can be quite large, and there can
+   be substantial duplication across items. The API allows the client to select
+   _which_ events are reported, but has no way to control which parts of a
+   matching event it wishes to receive.
+
+   This can be costly on the server (which has to marshal those data into
+   JSON), the network, and the client (which has to unmarshal the result and
+   then pick through for the components that are relevant to it).
+
+   Besides being inefficient, this also contributes to some of the persistent
+   connection issues mentioned above, e.g., filling up the websocket write
+   buffer and forcing the server to queue potentially several copies of a large
+   value in memory.
+
+5. **Client identity is tied to network address:** The Tendermint event API
+   identifies each subscriber by a (Client ID, Query) pair. In the RPC service,
+   the query is provided by the client, but the client ID is set to the TCP
+   address of the client (typically "host:port" or "ip:port").
+
+   This means that even if the server did _not_ drop subscriptions immediately
+   when the websocket connection is closed, a client may not be able to
+   reattach to its existing subscription.  Dialing a new connection is likely
+   to result in a different port (and, depending on their own proxy setup,
+   possibly a different public IP).
+
+   In isolation, this problem would be easy to work around with a new
+   subscription parameter, but it would require several other changes to the
+   handling of event subscriptions for that workaround to become useful.
+
+---
+## Decision
+
+(pending)
+
+To address the described problems, we will:
+
+1. Introduce a new API for event subscription to the Tendermint RPC service.
+   The proposed API is described in [Detailed Design](#detailed-design) below.
+
+2. This new API will target the Tendermint v0.36 release, during which the
+   current ("streaming") API will remain available as-is, but deprecated.
```

Comment:
From CosmJS's point of view you can remove the old API in 0.36 already. It is an optional feature for us anyways as we don't have it available when users give us HTTP RPC endpoints (i.e. the majority of cases).

Reply:
Sorry, I'm not sure I understand. The websocket _is_ exported via the HTTP RPC endpoint, it's routed for upgrade at `/websocket` on the same endpoint as the other methods. 

Are you saying you mostly don't export event subscriptions at all?

Reply:
I think this code illustrates our thinking:
1. If you give us a `wss?://`, we connect via websocket. If you give is a `https?://` we connect via the HTTP POST API: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/tendermint-rpc/src/tendermint34/tendermint34client.ts#L17-L24
2. So you end up with an `RpcClient` or an `RpcStreamingClient`: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/tendermint-rpc/src/rpcclients/rpcclient.ts#L28-L30
3. Don't try to use subscribe if you don't have a streaming client: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/tendermint-rpc/src/tendermint34/tendermint34client.ts#L335-L338
4. Or deal with the consequences: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/stargate/src/queries/queryclient.ts#L597-L610

Reply:
That being said, you are probably right that it can be implemented more dynamically and the HTTP -> WS switch can be done automatically. We just don't do it.

Reply:
Thanks for the references, that's very helpful. It sounds like—leaving aside the CORS issues you mentioned—having everything work correctly over HTTP could simplify most of this.

Reply:
Absolutely. The main point here is that we don't rely on subscriptions to work right now you you can break this if you want to. Happy to move to the new thing faster.

Reply:
This is all very useful feedback, I appreciate it. Do you have any insight about what other kinds of clients are using event subscription?

I hope we can agree to ship the first round of this in v0.36. I would feel safer not removing the existing API till the following release, but knowing that a substantial part of the userbase might not be afflicted by it is still valuable information.

Reply:
All other clients I see don't dare to interact with Tendermint RPC directly due to all the proto encoding/decoding necessary. The things I encounter in the wild use Tendermint RPC via CosmJS or the LCD API. However, this is  just one tiny perspective of a big a chaotic ecosystem.

Reply:
I'm talking about
1. the gRPC service that is accessed through Tendermint RPC for all the queries
2. the message/transaction encoding

All of this is needed for the payloads that then go to JSON. Lots of manual stuff that makes app developers still use the the LCD API.

Reply:
> 1. the gRPC service that is accessed through Tendermint RPC for all the queries

I think this must be something the SDK provides. TM core does not export any gRPC services.

> 2. the message/transaction encoding

Yes, the spelling of a transaction is mostly opaque to the consensus node, though I think as you say it's fairly common for apps to use protobuf here. For the TM RPC, the body of a transaction is just a blob of bytes, though.

> All of this is needed for the payloads that then go to JSON. Lots of manual stuff that makes app developers still use the the LCD API.

Sorry, that's not a term I'm familiar with. What do you mean by the "LCD API"?

Reply:
> I think this must be something the SDK provides. TM core does not export any gRPC services.

Oh it does via ABCI queries. I use this nice like 2019 before there was even gRPC:
- Encode query request to serialized proto data: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/stargate/src/queries/bank.ts#L23-L30
- Build paths: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/stargate/src/queries/utils.ts#L41-L44
- Send ABCI query: https://github.com/cosmos/cosmjs/blob/v0.27.0/packages/stargate/src/queries/queryclient.ts#L573-L585

> I think as you say it's fairly common for apps to use protobuf here. For the TM RPC, the body of a transaction is just a blob of bytes, though.

Right. Not too long ago there was Amino and nobody could encode Amino. The LCD API did this for the user on the server side. People are still used to the interface.

> Sorry, that's not a term I'm familiar with. What do you mean by the "LCD API"?

The API that the Cosmos Light Client Deamon offers. It is now integrated in the main process but it tried to preserve the same interfaces as in the old days. Many people call it "REST" but it has nothing to do with REST other than being JSON oder HTTP, so I refuse to use that term.

Reply:
> Oh it does via ABCI queries. I use this since like 2019 before there was gRPC even:

Oh yes, that's right, we removed that in #7121. I can see why that's an odd fit.

> The API that the Cosmos Light Client Deamon offers. It is now integrated in the main process AFAIK but it tried to preserve the same interfaces as in the old days.

Got it, thanks for the background!